### PR TITLE
[spirv] Add support for -fvk-bind-register

### PR DESCRIPTION
**Pull request by antiagainst**
format: `-fvk-bind-register <type-number> <space> <binding> <set>`

This option gives the ultimate manual control of descriptor
assignment. It requires:

* All resources are annotated with :register() in the source code
* -fvk-bind-register is specified for every resource

It overrules all other mechanisms.
It cannot be used together with -fvk-{u|b|s|t}-shift.

### CODE REVIEW

File: lib/DxcSupport/HLSLOptions.cpp [hunk 1]
```diff
@@ -206,6 +206,59 @@ static bool GetTargetVersionFromString(llvm::StringRef ref, unsigned *major, uns
   }
 }
 
+// SPIRV Change Starts
+#ifdef ENABLE_SPIRV_CODEGEN
+/// Checks and collects the arguments for -fvk-{b|s|t|u}-shift into *shifts.
+static bool handleVkShiftArgs(const InputArgList &args, OptSpecifier id,
```
Comment (ehsannas) [inline lib/DxcSupport/HLSLOptions.cpp]: why does this need to be `static`  ?

Comment (antiagainst) [reply lib/DxcSupport/HLSLOptions.cpp]: To keep this function local to this file.

**Comment by mikeh001** [comment]
There are potentially a large number of registers being set and the option name  -fvk-bind-register is quite long; this could lead to very long command lines compared to the equivalent option in glslang which is just --rsb. Would it be possible to have a shorter alias, or a way to set multiple registers with a single option?

**Comment by antiagainst** [comment]
@mikeh001: We are unlikely to copy what Glslang has into DXC; Glslang only has SPIR-V CodeGen, but DXC has both DXIL and SPIR-V. We'd like to make sure that command-line options for Vulkan/SPIR-V are distinguishable.  That's why we have the `-fvk`/`-fspv` prefix on Vulkan/SPIR-V related options. We also want the options to be self-explanatory, consistent among themselves and with existing compiler conventions. So I use `-bind-register` here to emphasize the mapping from a HLSL register to a Vulkan descriptor; hopefully this also gives enough hints regarding the option parameters.

As for taking multiple mapping relations using one instance of `-fvk-bind-register`, I think we have a trade-off here. I understand that it can potentially save several characters for each relation, but with multiple relations jammed together, it is very easy to miss a parameter or get something wrong. Besides, Clang command-line parsing library does not provide the flexibility of handling random-number parameters. Bypassing the existing command-line handling mechanism and providing a ad-hoc one for this option is really a hack that I would avoid. 

Yes, agreed this option can be invoked a lot of times on the command-line. So we are fine to create a short alias for it. How about `-vkbr`? It has the same number of characters as `--rsb`.

**Comment by mikeh001** [comment]
@antiagainst I totally agree with your reasoning, and -vkbr would work perfectly. Thanks!

**Comment by antiagainst** [comment]
One more thing to add is that the format we choose is different than Glslang's `--rsb`. Glslang's `--rsb` takes the following format:

```
--rsb <shader-stage> <register-type-number> <descriptor-set> <descriptor-binding>
```

I'm not sure how much value the `<shader-stage>` provides here; we are compiling for a certain entry point, whose shader stage is already determined by `-T`. And apparently register space is missing in the format. So instead we are doing:

```
-fvk-bind-register <type-number> <space> <binding> <set>
```
Not that to keep consistency with `:register(xX, spaceY)`, `[[vk::binding(X, Y)]`, and `-fvk-*-shift <shift> <space>`, we are placing Vulkan descriptor set number after binding number.

Let me know if you think this does not satisfy your needs or you think it can be improved. Thanks!

**Comment by ehsannas** [comment]
LGTM

**Pull request merged**